### PR TITLE
enterprise: add full audit log [AUTH-458]

### PR DESCRIPTION
**Pull request by BeryJu**
## Details

Audit logging that shows which fields were changed

needs:

- [x] Testing
- [ ] more sanitization
- [x] potentially replacing deepdiff with a simpler diff
- [x] Some way to ignore certain model fields being changed (`last_login` and `password` on user for example since they have dedicated events (`login` and `password_set`))

---

## Checklist

-   [ ] Local tests pass (`ak test authentik/`)
-   [ ] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ ] The code has been formatted (`make web`)
-   [ ] The translation files have been updated (`make i18n-extract`)

If applicable

-   [ ] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

### CODE REVIEW

File: authentik/enterprise/apps.py [hunk 1]
```diff
@@ -32,3 +32,9 @@ def check_enabled(self):
         from authentik.enterprise.models import LicenseKey
 
         return LicenseKey.get_total().is_valid()
+
+    def reconcile_install_middleware(self):
```
Comment (rissson) [inline authentik/enterprise/apps.py:36]:
```suggestion
    def reconcile_global_install_middleware(self):
```

File: authentik/enterprise/audit/apps.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+class AuthentikEnterpriseAuditConfig(EnterpriseConfig):
+    """Enterprise app config"""
+
+    name = "authentik.enterprise.audit"
+    label = "authentik_enterprise_audit"
+    verbose_name = "authentik Enterprise.Audit"
+    default = True
+
+    def reconcile_install_middleware(self):
```
Comment (rissson) [inline authentik/enterprise/audit/apps.py:15]:
```suggestion
    def reconcile_global_install_middleware(self):
```

File: authentik/enterprise/apps.py (hunk 1 above)
Comment (rissson) [reply authentik/enterprise/apps.py:36]: Also, I don't think this should be here, it only needs to be in the enterprise_audit app

File: authentik/enterprise/audit/apps.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+from django.conf import settings
+
+from authentik.enterprise.apps import EnterpriseConfig
+
+
+class AuthentikEnterpriseAuditConfig(EnterpriseConfig):
+    """Enterprise app config"""
+
+    name = "authentik.enterprise.audit"
+    label = "authentik_enterprise_audit"
```
Comment (rissson) [inline authentik/enterprise/audit/apps.py:11]: Did we want to stick with a label w/o `enterprise` in it in the event we move this to the OSS version?

File: authentik/enterprise/apps.py [hunk 4]
```diff
@@ -32,3 +32,9 @@ def check_enabled(self):
         from authentik.enterprise.models import LicenseKey
```
Comment (rissson) [inline authentik/enterprise/apps.py:32]: So above this I copied the `lru_cache` in the tenancy PR, but we should remove it

**Review by rissson — CHANGES_REQUESTED** [review]

File: authentik/enterprise/audit/apps.py (hunk 2 above)
Comment (rissson) [reply authentik/enterprise/audit/apps.py:15]: By always install the middleware, I meant adding it to settings.py (or apps.py which should overload it with the logic in settings.py)

Comment (BeryJu) [reply authentik/enterprise/audit/apps.py:15]: adding it to middleware in settings doesn't overwridte the main middleware list but rather just appends, so we have to do this here

Comment (rissson) [reply authentik/enterprise/audit/apps.py]: Ah yes, indeed, I missed that

**Review by rissson — APPROVED** [review]

**Pull request merged**